pkg/kubelet/kubelet.go

```diff
@@ -2003,14 +2003,27 @@ func (kl *Kubelet) SyncTerminatingPod(_ context.Context, pod *v1.Pod, podStatus
 
 	kl.probeManager.StopLivenessAndStartup(pod)
 
-	p := kubecontainer.ConvertPodStatusToRunningPod(kl.getRuntime().Type(), podStatus)
-	if err := kl.killPod(ctx, pod, p, gracePeriod); err != nil {
+	var podStopped bool
+	var err error
+	if !utilfeature.DefaultFeatureGate.Enabled(features.RestartContainerDuringTermination) {
+		p := kubecontainer.ConvertPodStatusToRunningPod(kl.getRuntime().Type(), podStatus)
+		err = kl.killPod(ctx, pod, p, gracePeriod)
+	} else {
+		podStopped, err = kl.syncTerminatingPod(ctx, pod, podStatus, gracePeriod, kl.getPullSecretsForPod(pod), kl.backOff)
+	}
+	if err != nil {
 		kl.recorder.Eventf(pod, v1.EventTypeWarning, events.FailedToKillPod, "error killing pod: %v", err)
 		// there was an error killing the pod, so we return that error directly
 		utilruntime.HandleError(err)
 		return err
 	}
 
+	if !podStopped {
+		// If the pod is not yet stopped, we should return an error to signal
+		// that the sync loop should back off.
+		return fmt.Errorf("pod is not yet stopped")
+	}
+
```

Comment:
/cc @smarterclayton 

https://github.com/kubernetes/enhancements/tree/master/keps/sig-node/4438-container-restart-termination
This PR is trying to change the (*kubelet).killPod in a non-blocking way so that other parts of the syncLoop can work during the container termination.

What do you think?